_posts/2017-04-04-what-every-developer-should-know-about-ir.md

```diff
@@ -0,0 +1,250 @@
+
+There has been loads of systems that use synonym dictionaries to add synonyms automatically to queries, so user don't have to think the synonyms at all. This is called *query expansion*, which is very important technique when helping user with queries. The problem is that you need a really good synonym dataset to make something like this automatically. This is usually really beneficial in large collections that deal with wide range of subjects - such as news collections - and is not that beneficial in smaller rather heavily specialised collections where users usually know the terms documents contain. Query expansion is not limited to boolean model, but usable in various other retrieval models as well, which we'll take a peek next.
+
+## Vector space model
+
+Vector space model is nowadays the most predominant retrieval model. I try not to get too mathematical, but in vector space model index is defined as a vector space as the name suggest. Now, this might seem like a total **mumbo-jumbo**, but one can think vector space as documents being vectors in space in which terms are dimensions (yes, there really is quite a lot of dimensions...) and weight of terms is "the juice" how far these vectors reach. Because vector calculus contains some very intriguing methods for calculating similarities we take advantage of that, and actually calculate the weighted similarity between query and a document and order result set with that. This is called *partial matching*, because documents that do not contain all the terms in query can easily be included to the result set. Interested reader is suggested to revise through suggested reading below, as we will not address the mathematical basis more here.
+
+It's important to notice that in pure vector space model, querying is effectively matching the similarity between the query and the documents. Hence, "give me more like this document" -functionality should be trivial to implement: the query is the document.
+
+### TW = TF * IDF
+
+This is the e = mc^2 of information retrieval, and can be written as: term weight equals **term frequency** times **inverse document frequency**.
+
+Most retrieval models choose the assumption that *if a term is common in document but rare in collection, it's really good to distinguish documents from each other within that collection*. So, if a document mentions lots of "dog", it really must be about dogs. It's especially important if "dog" is quite rare term in whole collection. Please note, that this assumption might be wrong as well within collection, but it still is one of the best assumptions we can make when dealing with texts in natural language. Think of it this way: would it make any sense to query with term "dog" to a collection of texts abouts dogs?
```

Comment:
common in _a_ document but rare in _the_ collection
_the_ documents from each other
is quite _a_ rare 
in _the_ whole collection
that this assumption might _as well_ be wrong within _the_ collection